### PR TITLE
*: fix autoid allocation to avoid quickly exhaustion

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

In https://github.com/pingcap/tidb/pull/13648, we try to allocate continuous row id for a single INSERT statement.

It asserts [the rows count](https://github.com/pingcap/tidb/pull/13648/files#diff-f92a6a9ee57cf92f8d63bd55d2edc05bR459) changed by a statement is known in advance, obtained through `stmtCtx.RecordRows()`.

However, that's not true.
In `insert into ... from select`, we don't know the actual rows count in advance.
`stmtCtx.RecordRows()` is accumulated in [`stmtCtx.AddRecordRows`](https://github.com/Little-Wallace/tidb/blob/f2844a1f8cfa0552050913e164fcb88bb7b4ed22/executor/insert.go#L64) in each chunk.
That means we count the rows repeatedly.

For example, `insert into ... from select` 100k rows, we assume the `rows` is known in advance and it's 100k.
However, each time the code run to `table.AddRecord`, the `rows` might be 

```
10k
20k
30k
...
80k
90k
100k
```

The total count is much more than 100k, which wastes a lot of auto-ID.

### What is changed and how it works?

I rewrite the `addRecord` function and add a reverse auto-ID hint parameter.
The caller would use the hint to guide the `table.AddRecord` to preallocate some auto-ID in the statement context.
And later use statement cached ones prior to the table's allocator.

### Check List 

Tests 

 - No code

Code changes

 - Has exported function/method change

### CODE REVIEW

File: table/tables/tables.go [hunk 1]
```diff
@@ -487,23 +487,19 @@ func (t *TableCommon) AddRecord(ctx sessionctx.Context, r []types.Datum, opts ..
...
-				}
-			}
-			stmtCtx.BaseRowID += 1
-			recordID = stmtCtx.BaseRowID
-		} else {
-			recordID, err = AllocHandle(ctx, t)
+		if opt.ReserveAutoID > 0 {
+			// Reserve a batch of auto ID in the statement ctx.
+			stmtCtx := ctx.GetSessionVars().StmtCtx
+			stmtCtx.BaseRowID, stmtCtx.MaxRowID, err = allocHandleIDs(ctx, t, uint64(opt.ReserveAutoID))
```
Comment (AilinKid) [inline table/tables/tables.go]: Maybe should comment here:
// actually, rows count won't be equal to the number allocated exactly, cause row may have `hasRecordID` when update.

File: table/tables/tables.go [hunk 2]
```diff
@@ -1019,6 +1015,14 @@ func GetColDefaultValue(ctx sessionctx.Context, col *table.Column, defaultVals [
 
 // AllocHandle allocate a new handle.
 func AllocHandle(ctx sessionctx.Context, t table.Table) (int64, error) {
+	if stmtCtx := ctx.GetSessionVars().StmtCtx; stmtCtx != nil {
```
Comment (AilinKid) [inline table/tables/tables.go]: Take care of some test, `ctx` maybe nil here, better check it.

Comment (tiancaiamao) [reply table/tables/tables.go]: Done

**Comment by AilinKid** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @Little-Wallace

**Comment by tiancaiamao** [comment]
PTAL @Little-Wallace

File: executor/insert.go [hunk 3]
```diff
@@ -81,8 +81,14 @@ func (e *InsertExec) exec(ctx context.Context, rows [][]types.Datum) error {
 			return err
 		}
 	} else {
-		for _, row := range rows {
-			if _, err := e.addRecord(ctx, row); err != nil {
+		for i, row := range rows {
+			var err error
+			if i == 0 {
+				_, err = e.addRecordWithAutoIDHint(ctx, row, len(rows))
```
Comment (Little-Wallace) [inline executor/insert.go]: Do we not need to limit the size of reserveAutoIDCount?

Comment (tiancaiamao) [reply executor/insert.go]: Unnecessary, the rows count equals to the size of a chunk https://github.com/pingcap/tidb/blob/master/executor/insert_common.go#L420

@Little-Wallace

**Comment by tiancaiamao** [comment]
Fix https://github.com/pingcap/tidb/issues/15073
Ping @Little-Wallace

**Pull request merged**